experimental/FTheoryTools/src/FamilyOfG4Fluxes/attributes.jl

```diff
@@ -94,3 +94,148 @@ true
+evaluate this polynomial for the numeric coefficient values that correspond to the given $G_4$-flux.
+
+We use symbols $a_i$ to indicate integral coefficients and $r_i$ to indicate rational coefficients.
+
+```jldoctest; setup = :(Oscar.LazyArtifacts.ensure_artifact_installed("QSMDB", Oscar.LazyArtifacts.find_artifacts_toml(Oscar.oscardir)))
+julia> qsm_model = literature_model(arxiv_id = "1903.00009", model_parameters = Dict("k" => 2021))
+Hypersurface model over a concrete base
+
+julia> fgs = special_flux_family(qsm_model, check = false)
+A family of G4 fluxes:
+  - Elementary quantization checks: satisfied
+  - Verticality checks: failed
+  - Non-abelian gauge group: broken
+
+julia> d3_tadpole_constraint(fgs);
+```
+"""
+function d3_tadpole_constraint(fgs::FamilyOfG4Fluxes; check::Bool = true)
+
+  # Is the result known?
+  if has_attribute(fgs, :d3_tadpole_constraint)
+    return get_attribute(fgs, :d3_tadpole_constraint)::QQMPolyRingElem
+  end
+
+  # Entry checks
+  m = model(fgs)
+  @req base_space(m) isa NormalToricVariety "Computation of D3-tadpole constraint only supported for toric base and ambient spaces"
+  @req dim(ambient_space(m)) == 5 "Computation of D3-tadpole constraint only supported for 5-dimensional toric ambient spaces"
+  if check
+    @req is_complete(ambient_space(m)) "Computation of D3-tadpole constraint only supported for complete toric ambient spaces"
+    @req is_simplicial(ambient_space(m)) "Computation of D3-tadpole constraint only supported for simplicial toric ambient space"
+  end
+
+  # Are intersection numbers known?
+  inter_dict = get_attribute!(m, :inter_dict) do
+    Dict{Tuple, ZZRingElem}()
+  end::Dict{Tuple, ZZRingElem}
+  s_inter_dict = get_attribute!(m, :s_inter_dict) do
+    Dict{String, ZZRingElem}()
+  end::Dict{String, ZZRingElem}
+  
+  # Compute data, that is frequently used by the sophisticated intersection product
+  #if arxiv_doi(m) == "10.48550/arXiv.1511.03209"
+    S = cox_ring(ambient_space(m))
+    #return S
+    gS = gens(cox_ring(ambient_space(m)))
+    linear_relations = matrix(QQ, matrix(ZZ, rays(ambient_space(m))))
+    scalings = [c.coeff for c in S.d]
+    mnf = Oscar._minimal_nonfaces(ambient_space(m))
+    sr_ideal_pos = [Vector{Int}(Polymake.row(mnf, i)) for i in 1:Polymake.nrows(mnf)]
+    data = (
+      S = S,
+      gS = gS,
+      linear_relations = linear_relations,
+      scalings = scalings,
+      sr_ideal_pos = sr_ideal_pos
+    )
+  #end
+
+  # Find the number of integral and rational parameters
+  numb_int_parameters = ncols(matrix_integral(fgs))
+  numb_rat_parameters = ncols(matrix_rational(fgs))
+
+  # Create a polynomial ring with parameters ai for the integral_parameters and ri for the rational parameters
+  gens_strings = vcat([["a$(k)" for k in 1:numb_int_parameters], ["r$(k)" for k in 1:numb_rat_parameters]]...)
+  amb_ring, gens = polynomial_ring(QQ, gens_strings)
```

Comment:
Could also do this
```suggestion
  amb_ring, _ = polynomial_ring(QQ, "a# => 1:numb_int_parameters, "r#" => 1: numb_rat_parameters)
  gens = gens(amb_ring)
```
Though better to rename the local variable `gens` -- it may mess up optimizations for calls to `gens` before it

Reply:
Even better to not throw away the gens in the first line and re-compute them in the second line